File: 0014-longest-common-prefix/0014-longest-common-prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        if (strs==null) return "";
        if (strs.length==1) return strs[0];
        
        int minLen = Integer.MAX_VALUE;
        for (String s : strs){
            if (s.length() < minLen){
                minLen = s.length();
            }
        }
        
        StringBuilder prefix = new StringBuilder();
        
        for (int i=0; i<minLen; i++){
            char curr = strs[0].charAt(i);

            for (int j=1; j<strs.length; j++){
                if (strs[j].charAt(i) != curr){
                    return prefix.toString();
                }
            }
            prefix.append(curr);
        }

        return prefix.toString();
    }
}

File: 0014-longest-common-prefix/0014-longest-common-prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        prefix = strs[0]
        
        for st in strs[1:]:
            # reduce prefix until it matches
            while not st.startswith(prefix):
                prefix = prefix[:-1]
                if prefix == "":
                    return ""
        return prefix